### PR TITLE
Simplify configuration options

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Resolves: #774

This pull request includes several changes aimed at improving the configuration and usage of options in the eBPF verifier code. The most important changes involve refactoring the way options are passed and used throughout the codebase, as well as simplifying the configuration structure.

### Refactoring and Simplification of Configuration Options:

* [`src/asm_cfg.cpp`](diffhunk://#diff-ab4aa4ed88c81d3f3ec54e8587af7e8a6b7a31059d4598f5869d640ce3fa3f83L362-R369): Changed the `prepare_cfg` function to use a `prepare_cfg_options` struct instead of multiple boolean parameters. [[1]](diffhunk://#diff-ab4aa4ed88c81d3f3ec54e8587af7e8a6b7a31059d4598f5869d640ce3fa3f83L362-R369) [[2]](diffhunk://#diff-ab4aa4ed88c81d3f3ec54e8587af7e8a6b7a31059d4598f5869d640ce3fa3f83L390-R388)
* [`src/config.hpp`](diffhunk://#diff-7d36a3982a659a42ce27f280729500523eb3cbe86ffdc80713f6b4aa83cc44d0R5-R24): Introduced the `prepare_cfg_options` struct and updated the `ebpf_verifier_options_t` struct to include this new struct. Removed default options initialization from `config.cpp`. [[1]](diffhunk://#diff-7d36a3982a659a42ce27f280729500523eb3cbe86ffdc80713f6b4aa83cc44d0R5-R24) [[2]](diffhunk://#diff-7d36a3982a659a42ce27f280729500523eb3cbe86ffdc80713f6b4aa83cc44d0L32)
* [`src/crab/cfg.hpp`](diffhunk://#diff-0f1616f116c0584e7ee0e0052bfee6e3e3537a6a993fdc45cf3c6fc794d35527L622-R631): Added the `prepare_cfg_options` struct and updated the `prepare_cfg` function signature accordingly.
* [`src/crab_verifier.cpp`](diffhunk://#diff-e3c65b3e2779bdd7a17c56ff48fd12ba98fcfcd19ab41081150358a9bbcdf920L165-R171): Updated various functions to use the new `ebpf_verifier_options_t` struct with the nested `prepare_cfg_options`. [[1]](diffhunk://#diff-e3c65b3e2779bdd7a17c56ff48fd12ba98fcfcd19ab41081150358a9bbcdf920L165-R171) [[2]](diffhunk://#diff-e3c65b3e2779bdd7a17c56ff48fd12ba98fcfcd19ab41081150358a9bbcdf920L186-R188) [[3]](diffhunk://#diff-e3c65b3e2779bdd7a17c56ff48fd12ba98fcfcd19ab41081150358a9bbcdf920L223-R221) [[4]](diffhunk://#diff-e3c65b3e2779bdd7a17c56ff48fd12ba98fcfcd19ab41081150358a9bbcdf920L239-R249)

### Consistent Usage of Configuration Options:

* [`src/asm_files.cpp`](diffhunk://#diff-a55f42b431b12b5af2a24cf20dc2ac08a2c8002f19f75bf73e6140b0bc46eaccL91-R91): Changed functions to pass `ebpf_verifier_options_t` by reference instead of pointer, ensuring consistency and simplifying the code. [[1]](diffhunk://#diff-a55f42b431b12b5af2a24cf20dc2ac08a2c8002f19f75bf73e6140b0bc46eaccL91-R91) [[2]](diffhunk://#diff-a55f42b431b12b5af2a24cf20dc2ac08a2c8002f19f75bf73e6140b0bc46eaccL119-R127) [[3]](diffhunk://#diff-a55f42b431b12b5af2a24cf20dc2ac08a2c8002f19f75bf73e6140b0bc46eaccL283-R282) [[4]](diffhunk://#diff-a55f42b431b12b5af2a24cf20dc2ac08a2c8002f19f75bf73e6140b0bc46eaccL314-R310)
* [`src/asm_files.hpp`](diffhunk://#diff-f7f8dd30e5bac013981cd163908011eb6f03554b096d742447bb5da9380ac3ccL14-R16): Updated function signatures to match the changes in `asm_files.cpp`.
* [`src/main/check.cpp`](diffhunk://#diff-11e7d5fe666d9226fee0e70f6c7ea915b975de07d0e69df3a3b17ae535c3a920L73-R73): Updated the main function to reflect changes in how options are passed and used, removing the dependency on default options. [[1]](diffhunk://#diff-11e7d5fe666d9226fee0e70f6c7ea915b975de07d0e69df3a3b17ae535c3a920L73-R73) [[2]](diffhunk://#diff-11e7d5fe666d9226fee0e70f6c7ea915b975de07d0e69df3a3b17ae535c3a920L100-R100) [[3]](diffhunk://#diff-11e7d5fe666d9226fee0e70f6c7ea915b975de07d0e69df3a3b17ae535c3a920L196-R196) [[4]](diffhunk://#diff-11e7d5fe666d9226fee0e70f6c7ea915b975de07d0e69df3a3b17ae535c3a920L211-R211) [[5]](diffhunk://#diff-11e7d5fe666d9226fee0e70f6c7ea915b975de07d0e69df3a3b17ae535c3a920L238-R240)

### Removal of Default Options Initialization:

* [`src/config.cpp`](diffhunk://#diff-6b2f0a449fdefd8930e23ef0dcd752beec69242e1303d77653f047c5e0766385L1-L17): Removed the initialization of default options, as the new approach does not rely on a global default options instance.

These changes collectively improve the maintainability and readability of the code by centralizing configuration options and ensuring consistent usage throughout the codebase.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new `prepare_cfg_options` structure to simplify configuration parameter handling in the control-flow graph preparation process.
	- Enhanced the handling of options in various functions for improved clarity and maintainability.
	- Added a new option, `must_have_exit`, to allow test cases to omit an exit instruction.

- **Bug Fixes**
	- Improved clarity in the handling of configuration options, which may affect behavior in certain scenarios.

- **Tests**
	- Updated test cases to align with the new configuration structure, ensuring comprehensive testing of eBPF program verification.
	- Modified test implementations to utilize empty initializer lists for parameter handling, enhancing consistency across tests.

### CODE REVIEW

File: src/crab/cfg.hpp [hunk 1]
```diff
@@ -619,8 +619,13 @@ std::vector<std::string> stats_headers();
...
 
-cfg_t prepare_cfg(const InstructionSeq& prog, const program_info& info, bool simplify, bool check_for_termination,
-                  bool must_have_exit = true);
+struct prepare_cfg_options {
+    bool simplify = true;
+    bool check_for_termination = true;
+    bool must_have_exit = true;
+};
+
+cfg_t prepare_cfg(const InstructionSeq& prog, const program_info& info, const prepare_cfg_options& options);
```
Comment (elazarg) [inline src/crab/cfg.hpp:628]: I don't remember - is it not possible to set default `options=prepare_cfg_options{}`?

File: src/crab/cfg.hpp [hunk 2]
```diff
@@ -619,8 +619,13 @@ std::vector<std::string> stats_headers();
 
 std::map<std::string, int> collect_stats(const cfg_t&);
 
-cfg_t prepare_cfg(const InstructionSeq& prog, const program_info& info, bool simplify, bool check_for_termination,
-                  bool must_have_exit = true);
+struct prepare_cfg_options {
```
Comment (elazarg) [inline src/crab/cfg.hpp:622]: Does it make sense to use this struct as a field in the ebpf_verifier_options_t?

Comment (Alan-Jowett) [reply src/crab/cfg.hpp:622]: Good idea. Let me see if I can make this change.

**Comment by Alan-Jowett** [comment]
@coderabbitai full review

File: src/crab/cfg.hpp [hunk 3]
```diff
@@ -614,8 +614,16 @@ std::vector<std::string> stats_headers();
...
+    /// When true, simplifies the control flow graph by merging basic blocks.
+    bool simplify = true;
+    /// When true, verifies that the program terminates.
+    bool check_for_termination = false;
+    /// When true, ensures the program has a valid exit block.
+    bool must_have_exit = true;
+};
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/crab/cfg.hpp:624]: I don't understand this comment. prepare_cfg_options and prepare_cfg are collacated.

File: src/asm_files.hpp [hunk 4]
```diff
@@ -11,9 +11,9 @@
 
 std::vector<raw_program> read_raw(std::string path, program_info info);
 std::vector<raw_program> read_elf(const std::string& path, const std::string& desired_section,
-                                  const ebpf_verifier_options_t* options, const ebpf_platform_t* platform);
+                                  const ebpf_verifier_options_t& options, const ebpf_platform_t* platform);
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/asm_files.hpp:14]: The parameter options is a reference and hence can't be null. This comment makes no sense.

File: src/crab_verifier.hpp [hunk 5]
```diff
@@ -7,11 +7,11 @@
 #include "spec_type_descriptors.hpp"
 #include "string_constraints.hpp"
 
-bool run_ebpf_analysis(std::ostream& s, const cfg_t& cfg, const program_info& info, const ebpf_verifier_options_t* options,
-                       ebpf_verifier_stats_t* stats);
+bool run_ebpf_analysis(std::ostream& s, const cfg_t& cfg, const program_info& info,
+                       const ebpf_verifier_options_t& options, ebpf_verifier_stats_t* stats);
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/crab_verifier.hpp:11]: Stats is an optional out parameter. Some callers don't care about stats so gathering them would be redundant.

File: src/main/check.cpp [hunk 6]
```diff
@@ -124,7 +124,7 @@ int main(int argc, char** argv) {
         ->expected(0, _conformance_groups.size())
         ->check(CLI::IsMember(_get_conformance_group_names()));
 
-    app.add_flag("--simplify,!--no-simplify", ebpf_verifier_options.simplify,
+    app.add_flag("--simplify,!--no-simplify", ebpf_verifier_options.cfg_opts.simplify,
                  "Simplify the CFG before analysis by merging chains of instructions into a single basic block. "
                  "Default: enabled")
         ->group("Verbosity");
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/main/check.cpp:130]: Simplify doesn't affect the result of the analysis, instead controls how much additional information is available to understand the result of the analysis, hence verbosity is the correct group.

File: src/main/check.cpp [hunk 7]
```diff
@@ -235,9 +235,9 @@
     if (domain == "zoneCrab") {
         ebpf_verifier_stats_t verifier_stats;
         const auto [res, seconds] = timed_execution([&] {
-            return ebpf_verify_program(std::cout, prog, raw_prog.info, &ebpf_verifier_options, &verifier_stats);
+            return ebpf_verify_program(std::cout, prog, raw_prog.info, ebpf_verifier_options, &verifier_stats);
         });
-        if (res && ebpf_verifier_options.check_termination &&
+        if (res && ebpf_verifier_options.cfg_opts.check_for_termination &&
             (ebpf_verifier_options.print_failures || ebpf_verifier_options.print_invariants)) {
             std::cout << "Program terminates within " << verifier_stats.max_loop_count << " loop iterations\n";
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply src/main/check.cpp:242]: Disagree. It's clear enough as it is.

**Comment by Alan-Jowett** [comment]
> @Alan-Jowett sorry about the conflicts

No worries.

**Pull request merged**